taskmanager/taskq.go

```
@@ -89,76 +96,88 @@ func (c *Client) loadTaskQ() error {
+	c.options.logger.Info().Msgf("executing task: %s", options.TaskName)
 
-		return
+	// Try to get the task
+	task, ok := c.options.taskq.tasks[options.TaskName]
+	if !ok {
+		return fmt.Errorf("task %s not registered", options.TaskName)
 	}
 
-	// Register and store the task
-	c.options.taskq.tasks[task.Name] = taskq.RegisterTask(convertTaskToTaskQ(task))
+	// Task message will be used to add to the queue
+	taskMessage := task.WithArgs(ctx, options.Arguments...)
 
-	mutex.Unlock()
+	// There are two ways to run a task:
+	// Option 1: Run the task immediately
+	if options.RunEveryPeriod == 0 {
+		return c.options.taskq.queue.Add(taskMessage)
+	}
 
-	// Debugging
-	c.DebugLog(fmt.Sprintf("registering task: %s...", c.options.taskq.tasks[task.Name].Name()))
+	// Option 2: Run the task periodically using cron
+	// Note: The first run will be after the period has passed
+	return c.scheduleTaskWithCron(ctx, task, taskMessage, options.RunEveryPeriod)
 }
 
-// runTaskUsingTaskQ will run a task using TaskQ
-func (c *Client) runTaskUsingTaskQ(ctx context.Context, options *TaskOptions) error {
+func (c *TaskManager) scheduleTaskWithCron(ctx context.Context, task *taskq.Task, taskMessage *taskq.Message, runEveryPeriod time.Duration) error {
+	// When using Redis, we need to use a distributed timed lock to prevent the addition of the same task to the queue by multiple instances.
+	// With this approach, only one instance will add the task to the queue within a given period.
+	var tryLock func() bool
+	if c.Factory() == FactoryRedis {
+		key := fmt.Sprintf("taskq_cronlock_%s", task.Name())
 
-	// Starting the execution of the task
-	c.DebugLog(fmt.Sprintf(
-		"executing task: %s... delay: %s arguments: %s",
-		options.TaskName,
-		options.Delay.String(),
-		fmt.Sprintf("%+v", options.Arguments),
-	))
+		// The runEveryPeriod should be greater than 1 second
+		if runEveryPeriod < 1*time.Second {
+			runEveryPeriod = 1 * time.Second
```

maybe instead of default we could return an error here

---

Done.